collect_app/src/main/java/org/odk/collect/android/adapters/FormDownloadListAdapter.java

```diff
@@ -0,0 +1,97 @@
+import android.support.annotation.NonNull;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ArrayAdapter;
+import android.widget.CheckBox;
+import android.widget.LinearLayout;
+import android.widget.TextView;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.dao.FormsDao;
+import org.odk.collect.android.logic.FormDetails;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import static org.odk.collect.android.activities.FormDownloadList.FORMID_DISPLAY;
+import static org.odk.collect.android.activities.FormDownloadList.FORMNAME;
+import static org.odk.collect.android.activities.FormDownloadList.FORM_ID_KEY;
+
+public class FormDownloadListAdapter extends ArrayAdapter {
+
+    private ArrayList<HashMap<String, String>> filteredFormList;
+    private HashMap<String, FormDetails> formNamesAndURLs;
+
+    public FormDownloadListAdapter(Context context, ArrayList<HashMap<String, String>> filteredFormList, HashMap<String, FormDetails> formNamesAndURLs) {
+        super(context, R.layout.two_item_multiple_choice, filteredFormList);
+        this.filteredFormList = filteredFormList;
+        this.formNamesAndURLs = formNamesAndURLs;
+    }
+
+    private class ViewHolder {
+        TextView text1;
+        TextView text2;
+        LinearLayout updateWarning;
+        CheckBox checkBox;
+    }
+
+    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
+        View row = convertView;
+        final ViewHolder holder;
+        if (row == null) {
+            holder = new ViewHolder();
+            LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+            row = inflater.inflate(R.layout.two_item_multiple_choice, parent, false);
+
+            holder.text1 = (TextView) row.findViewById(R.id.text1);
+            holder.text2 = (TextView) row.findViewById(R.id.text2);
+            holder.updateWarning = (LinearLayout) row.findViewById(R.id.update_warning);
+            holder.checkBox = (CheckBox) row.findViewById(R.id.checkbox);
+            row.setTag(holder);
+        } else {
+            holder = (ViewHolder) row.getTag();
+        }
+
+        holder.text1.setText(filteredFormList.get(position).get(FORMNAME));
+        holder.text2.setText(filteredFormList.get(position).get(FORMID_DISPLAY));
+
+        if (isNewerFormAvailable(position)) {
+            holder.updateWarning.setVisibility(View.VISIBLE);
+        } else {
+            holder.updateWarning.setVisibility(View.GONE);
+        }
+
+        return row;
+    }
+
+    private boolean isNewerFormAvailable(int position) {
+        String formId = filteredFormList.get(position).get(FORM_ID_KEY);
+        String formVersion = formNamesAndURLs.get(formId).formVersion;
+        if (formVersion != null) {
+            Integer formVersionFormDatabase = new FormsDao().getFormVersionForFormId(formId);
```

Comment:
Should this be `formVersionFromDatabase`? (from rather than form)

Reply:
Right, fixed.